### PR TITLE
Fix jsonpath slice step handling

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fixes step handling of jsonpath expressions.

**Does this PR introduce a user-facing change?**:
```release-note
jsonpath expressions containing `[start:end:step]` slice are now evaluated correctly
```

/sig api-machinery
/cc @smarterclayton

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go [hunk 1]
```diff
@@ -519,3 +522,112 @@ func TestNegativeIndex(t *testing.T) {
...
+			},
+			{
+				"test containers[0:6:5]",
+				`{.spec.containers[0:6:5].name}`,
+				data,
+				"fake0 fake5",
+				false,
+			},
+			{
+				"test containers[0:6:6]",
```
Comment (smarterclayton) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go]: test part of the slice?  [0:3], [0:3:3]

Comment (liggitt) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go]: added

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go [hunk 2]
```diff
@@ -273,12 +273,16 @@ func (j *JSONPath) evalArray(input []reflect.Value, node *ArrayNode) ([]reflect.
...
-		if !params[2].Known {
-			value = value.Slice(params[0].Value, params[1].Value)
-		} else {
-			value = value.Slice3(params[0].Value, params[1].Value, params[2].Value)
+		value = value.Slice(params[0].Value, params[1].Value)
+
+		step := 1
+		if params[2].Known {
+			if params[2].Value <= 0 {
+				return input, fmt.Errorf("step must be >= 0")
```
Comment (tedyu) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: A bit curious: why is step of zero allowed ?

Comment (liggitt) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: it is not allowed. a step size of 0 means that no progress would be made between the start and end

Comment (liggitt) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: oh, I see... the message is incorrect

Comment (tedyu) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: Should I send out PR to correct the message ?

Comment (liggitt) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: sure, thanks

Comment (tedyu) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: Created #73179

**Pull request merged**